Source/Core/VideoBackends/D3D12/D3DTexture.cpp

```diff
@@ -0,0 +1,261 @@
+	};
+
+	ResourceBarrier(currentCommandList, pTexture12, currentResourceState, D3D12_RESOURCE_STATE_COPY_DEST, level);
+
+	CHECK(0 != UpdateSubresources(currentCommandList, pTexture12, s_replaceRGBATexture2DUploadHeap, s_replaceRGBATexture2DUploadHeapOffset, level, 1, &subresourceData12), "UpdateSubresources failed.");
+
+	ResourceBarrier(D3D::currentCommandList, pTexture12, D3D12_RESOURCE_STATE_COPY_DEST, currentResourceState, level);
+
+	s_replaceRGBATexture2DUploadHeapOffset += uploadSize;
+
+	s_replaceRGBATexture2DUploadHeapOffset = (s_replaceRGBATexture2DUploadHeapOffset + D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT - 1) & ~(D3D12_TEXTURE_DATA_PLACEMENT_ALIGNMENT - 1); // Align offset in upload heap to 512 bytes, as required.
+}
+
+}  // namespace
+
+D3DTexture2D* D3DTexture2D::Create(unsigned int width, unsigned int height, D3D11_BIND_FLAG bind, D3D11_USAGE usage, DXGI_FORMAT fmt, unsigned int levels, unsigned int slices, D3D12_SUBRESOURCE_DATA* data)
+{
+	ID3D12Resource* pTexture12 = nullptr;
+
+	D3D12_RESOURCE_DESC texdesc12 = CD3DX12_RESOURCE_DESC::Tex2D(
+		fmt,
+		width,
+		height,
+		slices,
+		levels
+		);
+
+	D3D12_CLEAR_VALUE optimizedClearValue = {};
+	optimizedClearValue.Format = fmt;
+
+	if (bind & D3D11_BIND_RENDER_TARGET)
+	{
+		texdesc12.Flags |= D3D12_RESOURCE_FLAG_ALLOW_RENDER_TARGET;
+		optimizedClearValue.Color[0] = 0.0f;
+		optimizedClearValue.Color[1] = 0.0f;
+		optimizedClearValue.Color[2] = 0.0f;
+		optimizedClearValue.Color[3] = 1.0f;
+	}
+
+	if (bind & D3D11_BIND_DEPTH_STENCIL)
+	{
+		texdesc12.Flags |= D3D12_RESOURCE_FLAG_ALLOW_DEPTH_STENCIL;
+		optimizedClearValue.DepthStencil.Depth = 0.0f;
+		optimizedClearValue.DepthStencil.Stencil = 0;
+	}
+
+	CheckHR(
+		D3D::device12->CreateCommittedResource(
+			&CD3DX12_HEAP_PROPERTIES(D3D12_HEAP_TYPE_DEFAULT),
+			D3D12_HEAP_FLAG_NONE,
+			&CD3DX12_RESOURCE_DESC(texdesc12),
+			D3D12_RESOURCE_STATE_PIXEL_SHADER_RESOURCE,
+			&optimizedClearValue,
+			IID_PPV_ARGS(&pTexture12)
+			)
+		);
+
+	D3D::SetDebugObjectName12(pTexture12, "Texture created via D3DTexture2D::Create");
+	D3DTexture2D* ret = new D3DTexture2D(pTexture12, bind, DXGI_FORMAT_UNKNOWN, DXGI_FORMAT_UNKNOWN, DXGI_FORMAT_UNKNOWN, false, D3D12_RESOURCE_STATE_PIXEL_SHADER_RESOURCE);
+
+	if (data)
+	{
+		DX12::D3D::ReplaceRGBATexture2D(pTexture12, reinterpret_cast<const u8*>(data->pData), width, height, static_cast<unsigned int>(data->RowPitch), 0, D3D12_RESOURCE_STATE_PIXEL_SHADER_RESOURCE);
+	}
+
+	SAFE_RELEASE(pTexture12);
+	return ret;
+}
+
+void D3DTexture2D::AddRef()
+{
+	++ref;
+}
+
+UINT D3DTexture2D::Release()
+{
+	--ref;
+	if (ref == 0)
+	{
+		delete this;
+		return 0;
+	}
+	return ref;
+}
+
+ID3D12Resource* &D3DTexture2D::GetTex12() { return tex12; }
+D3D12_CPU_DESCRIPTOR_HANDLE &D3DTexture2D::GetSRV12cpu() { return srv12cpu; }
+D3D12_GPU_DESCRIPTOR_HANDLE &D3DTexture2D::GetSRV12gpu() { return srv12gpu; }
+D3D12_CPU_DESCRIPTOR_HANDLE &D3DTexture2D::GetSRV12gpuCpuShadow() { return srv12gpuCpuShadow; }
```

Comment:
Names should have varying case like this. Initialisms should be uppercased

e.g.

```cpp
GetSRV12GPUCPUShadow
```